### PR TITLE
Make login form accessible to Password Managers

### PR DESCRIPTION
**Pull request by parasharrajat**
### Fixed Issues

$ https://github.com/Expensify/App/issues/3855
$ https://github.com/Expensify/App/issues/5956

### Tests | QA Steps
1. Navigate to NewDot on the web.
2. Right-click on the email field.
3. Click on the 1Password option to show the autofill.
4. 1Password Autofill option should be shown on the input.

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

#### Web

Updating Shortly...

### CODE REVIEW

**Comment by AndrewGable** [comment]
Conflicts!

File: src/libs/TextInputUtils/index.js
Comment (marcaaron) [inline src/libs/TextInputUtils/index.js:1]: Needs a JS doc

File: src/libs/TextInputUtils/index.js
Comment (marcaaron) [inline src/libs/TextInputUtils/index.js:6]: Please add to the comment an explanation of why this is needed.

File: src/libs/TextInputUtils/index.js
Comment (marcaaron) [inline src/libs/TextInputUtils/index.js:13]: Please use an early return here when there is no value for `input`

File: src/libs/TextInputUtils/index.native.js
Comment (marcaaron) [inline src/libs/TextInputUtils/index.native.js:1]: JS Doc

File: src/libs/TextInputUtils/index.native.js
Comment (marcaaron) [inline src/libs/TextInputUtils/index.native.js:5]: One reason why we move things into `index.js` and `index.native.js` files is so we don't have to have code that mentions a specific platform. Can we name this something more generic and also add some explanation about why we are not running this code on native?

File: src/libs/TextInputUtils/index.native.js
Comment (marcaaron) [inline src/libs/TextInputUtils/index.native.js:8]: Can we just return a string constant instead so we don't have to call a function?

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [inline src/pages/signin/PasswordForm.js:104]: Just curious but what do we need to set the `nativeID` for ?

File: src/pages/signin/SignInPageLayout/SignInPageLayoutNarrow.js
Comment (marcaaron) [inline src/pages/signin/SignInPageLayout/SignInPageLayoutNarrow.js:31]: early return please

File: src/pages/signin/SignInPageLayout/SignInPageLayoutNarrow.js
Comment (marcaaron) [inline src/pages/signin/SignInPageLayout/SignInPageLayoutNarrow.js:33]: We should encapsulate this logic in a new kind of view then pass these as props. The component can then handle setting these native props with this `componentDidMount()` logic instead of having to repeat it everywhere we want to do this.

Comment (marcaaron) [reply src/pages/signin/SignInPageLayout/SignInPageLayoutNarrow.js:33]: Also please add some comments about why `method` and `action` are required in this case

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (marcaaron) [inline src/components/ExpensiTextInput/BaseExpensiTextInput.js:48]: If there is no input we should return early instead of checking on like 48 and 45

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Comment by parasharrajat** [comment]
Thanks for the review Marc. I will resolve them as soon as possible.

File: src/libs/TextInputUtils/index.native.js
Comment (parasharrajat) [reply src/libs/TextInputUtils/index.native.js:5]: Ok. I will update this.
Mostly, the reason is that I am using this to set the web attributes needed for browser perspective. (name, for, ...etc).

File: src/libs/TextInputUtils/index.native.js
Comment (parasharrajat) [reply src/libs/TextInputUtils/index.native.js:8]: Yeah. Sure.

File: src/pages/signin/PasswordForm.js
Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js:104]: For Attaching the label with input. This could be optional. There are a lot of uncertainties in the case of password Managers so I used everything that we generally do for login forms. Another reason was to just point out these things in the review and decide if they are necessary.

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (parasharrajat) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:48]: Yes, absolutely.

**Comment by marcaaron** [comment]
What's the status on this one @parasharrajat ?

**Comment by parasharrajat** [comment]
@rushatgabhane Could you please review this and tell me if it works for you?

**Comment by rushatgabhane** [comment]
> Ideally, once 1P is used for entering the email address, the Password on the following page will be autofilled as well.

I tested this branch with Bitwarden as my password manager.
The password field in next page isn't autofilled.

https://user-images.githubusercontent.com/29673073/140638533-7d600f5f-0bac-427f-ad7a-710c0681ef3e.mp4

**Comment by rushatgabhane** [comment]
For Android, I'm getting this error on launch.


```
Warning: Failed prop type: ExpensiTextInputLabel: prop type `for` is invalid; it must be a function, usually from the `prop-types` package, but received `string`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.
    in ExpensiTextInputLabel (at BaseExpensiTextInput.js:189)
    in RCTView (at View.js:34)
    in View (at BaseExpensiTextInput.js:177)
    in TouchableWithoutFeedback (at BaseExpensiTextInput.js:176)
    in RCTView (at View.js:34)
    in View (at BaseExpensiTextInput.js:170)
    in RCTView (at View.js:34)
    in View (at BaseExpensiTextInput.js:169)
    in BaseExpensiTextInput (at ExpensiTextInput/index.android.js:7)

```

**Comment by parasharrajat** [comment]
@rushatgabhane Thanks for testing it. I have pushed new changes, please check again. Let's see I tested with lastpass.

**Comment by rushatgabhane** [comment]
@parasharrajat Works perfectly! 💯

**Comment by parasharrajat** [comment]
Thanks, @rushatgabhane. It would have not been possible without your help.

If you are curious to know, what was the issue?

1. Password Managers do not track dom changes unless page changes. So when we were switching between the login field component to the password field component, the selection of login you made in the login page from the password manager is also lost during this transition.

To fix this, we have to keep all the form fields, login, and password in the dom, keeping the relation between them. 

Also, as we use the multistep form, I have to visually hide the other dom nodes based on the current step to keep the UI intact.

**Comment by rushatgabhane** [comment]
Thanks for the clarity. It makes sense now :)

**Comment by parasharrajat** [comment]
I will soon finalize the PR. All changes are done. Cleanup is left.

File: src/pages/signin/PasswordForm.js
Comment (rushatgabhane) [inline src/pages/signin/PasswordForm.js:50]: Please remove all refs set as null in constructor.
[Source](https://github.com/Expensify/App/pull/2143)

**Comment by parasharrajat** [comment]
@marcaaron This is ready.

File: src/pages/signin/SignInPage.js
Comment (parasharrajat) [inline src/pages/signin/SignInPage.js:91]: This is change is necessary. Most password managers do not work with dynamically added fields. So When the user selected the username in LoginForm and navigate to PasswordForm that autofill will be lost and he has to refill the value from suggestions.

To make it work with them, I just hid them behind the CSS and kept them in DOM. But I think LoginForm and PasswordForm are not needed or not part of `ResendValidationForm` so these should be unmounted for `showResendValidationLinkForm`.

I can't use `{!showResendValidationLinkForm  &&  <> ...</>}` this reattach those forms and break it. If I extract them into another PureComponent then return null for `!showResendValidationLinkForm ` it will work. I need your suggestion here. Thanks.

**Review by rushatgabhane — CHANGES_REQUESTED** [review]

File: src/pages/signin/SignInPage.js
Comment (parasharrajat) [reply src/pages/signin/SignInPage.js:91]: This will solve https://github.com/Expensify/App/issues/5956 as well.

File: src/pages/signin/PasswordForm.js
Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js:50]: Maybe it's about adding them. :smile:

Comment (rushatgabhane) [reply src/pages/signin/PasswordForm.js:50]: No no, the PR was reverted.
And we agreed to clean it up! [Slack ref](https://expensify.slack.com/archives/C01GTK53T8Q/p1635178933076400?thread_ts=1634850853.025000&cid=C01GTK53T8Q)

Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js:50]: Thanks. I am still not sure but I will check the styleGuide before doing it or maybe wait for the review to combine changes into one.

**Comment by parasharrajat** [comment]
This is ready.

**Comment by parasharrajat** [comment]
Hey, @marcaaron. Please review it. This could sound hacky too but I think it's not based on the following reasons.
1. I tried to add proper HTML syntax so that browser understands our form.
2. Using a visually hidden technique, will remove the need to fill the password manually from password managers.
3. Password manager will only autofill if input fields are present on the page afterload (NO JS magic). 

I improved the code from the first edition but I sure can be improved.

File: src/components/Form.js
Comment (Jag96) [inline src/components/Form.js:24]: Unnecessary displayName here

File: src/components/Form.js
Comment (Jag96) [inline src/components/Form.js:8]: Is this still required? If so can you add a comment why?

File: src/components/ExpensiTextInput/baseExpensiTextInputPropTypes.js
Comment (Jag96) [inline src/components/ExpensiTextInput/baseExpensiTextInputPropTypes.js:7]: Since this is only used on web, should this just be added in index.js instead of including it here?

**Review by Jag96 — COMMENTED** [review]
Tests well on web, just a couple comments. Also, it looks like this does fix https://github.com/Expensify/App/issues/5956 so updating description for that 🎉

File: src/components/ExpensiTextInput/baseExpensiTextInputPropTypes.js
Comment (parasharrajat) [reply src/components/ExpensiTextInput/baseExpensiTextInputPropTypes.js:7]: Even though it is used for the Web, we will have to pass it to `ExpensiTextInput` irrespective of plaform. It has to be passed down to the BaseExpensiTextInput. we can simply pass null as the default value for native but still, it needs to be passed.
so prop definition is necessary. Just it does not affect non-web platforms.

File: src/components/Form.js
Comment (parasharrajat) [reply src/components/Form.js:8]: I don't know thee reason but it is mentioned on each Password Manager to add these attributes and this is Web standard.

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [inline src/pages/signin/PasswordForm.js:37]: Same comment about using an `is*` prefix

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [reply src/pages/signin/PasswordForm.js:50]: +1

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [inline src/pages/signin/PasswordForm.js:72]: NAB, this logic looks very similar to the logic in `LoginForm`. I'm not sure how common this pattern will be after this PR, but maybe there is some way to abstract it (instead of copy paste everywhere this trick is needed).

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [inline src/pages/signin/LoginForm.js:134]: Should we create a constant fo this as well?

File: src/pages/signin/SignInPage.js
Comment (marcaaron) [reply src/pages/signin/SignInPage.js:91]: >When the user selected the username in LoginForm and navigate to PasswordForm that autofill will be lost and he has to refill the value from suggestions

Which autofill will be lost? If we navigate to the password form the login form value is lost?

Not following the rest of the proposal, sorry. But I think we should avoid manipulating any CSS directly.

**Review by marcaaron — CHANGES_REQUESTED** [review]
Left a few comments. I think the main blockers for me are:

- I'm not sure why we are doing the `visible` change instead of conditionally rendering the forms. If I came across the code I'd be completely confused about why we are doing things this way.
- `InputUtils.setBrowserAttributes()` feels like the wrong way to pass attributes to non-native components. I might be convinced otherwise, but if we have non-native components that need props native components do not - then we should create two different versions of that component.

File: src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js
Comment (parasharrajat) [reply src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js:14]: Ok, I am just a little confused about achieving this. TextInput needs this prop which is used in BaseExpensiInput. 
Are you suggesting that
1. I make a component TextinputWrapper, native, web.
2. Then in web one, use setNativeProps to set it.

But still name prop needs to be passed through ExpensiInput all the way to web version on Web.

File: src/pages/signin/LoginForm.js
Comment (parasharrajat) [reply src/pages/signin/LoginForm.js:79]: 1. Password Managers parse all the inputs that are rendered together in a form.
2. Now on the email input you chose en email for prefilling.
3. If the password field is there it will fill the password.
4. You use js to swap the input field with the password field, the Password manager's autofill state is lost. You have to again fill in the data from the password manager. (it will need the master password again). https://github.com/Expensify/App/issues/5956.

5. On the first render, If you don't have password input, it will not be autofilled. (Google Chrome works but LastPass , 1Pass does not).

File: src/pages/signin/LoginForm.js
Comment (parasharrajat) [reply src/pages/signin/LoginForm.js:134]: Does not seem necessary as it should be probably be used in one place. But we can.

File: src/pages/signin/PasswordForm.js
Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js:72]: Yeah, good Idea

File: src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js
Comment (marcaaron) [reply src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js:14]: >I make a component TextinputWrapper, native, web.

Not a wrapper. I was thinking to just pass the prop down, but only use it in the non native versions. We don't need to mention here that "only web" uses this. We should just make it so all `ExpensiTextInput` get the generic `for` prop, but native doesn't use it. Make sense?

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [reply src/pages/signin/LoginForm.js:79]: Ok thanks! Sorry, I am a little slow to understand sometimes 😅 

I'm not really seeing an alternative solution. 

But I think we should add some comments to the code to explain why we are changing the visibility in `SignInPage` and not conditionally rendering. Otherwise, someone may later decide that they prefer conditional rendering and unintentionally break the password manager.

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [reply src/pages/signin/LoginForm.js:134]: Good point. it's a NAB for me in that case - I'll leave it up to you to decide.

File: src/pages/signin/LoginForm.js
Comment (parasharrajat) [reply src/pages/signin/LoginForm.js:79]: Yeah, I would abstract this into a component so that we can use it. and comment there.

File: src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js
Comment (parasharrajat) [reply src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js:14]: Actually, it does that. `for` is passed as `' '` for native. As `name` prop is not available on the RN TextInput we have to use `setNativeProps` so I thought to just noop  `setNativeProps` on native. 

I used "only web" so that we know it does not have any effect on native platforms. 

Thus I created `setBrowserAttribute`.

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (parasharrajat) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:50]: > The problem with the way we have this is that someone might look at the underlying method and think they can use it to setNativeProps but it just noops on native.

Good point. I think I can just pass an empty string for the name prop from .native.js file and pass the proper name prop down to this from index.js file.

And use setNativeProps here.

File: src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js
Comment (marcaaron) [reply src/components/ExpensiTextInput/ExpensiTextInputLabel/expensiTextInputLabelPropTypes.js:14]: Yeah, I think we should not have a method that refers to a specific platform that can be used anywhere. Better to set the native props for whatever the component in `index.js` or `index.web.js` files etc and have the `index.native.js` file do nothing with the prop.

**Comment by parasharrajat** [comment]
@marcaaron Ready for review. Major refactor... :smile:

File: src/components/CollapsibleView.js
Comment (marcaaron) [inline src/components/CollapsibleView.js:18]: `ToggleVisibilityView` maybe? Collapsible makes me think of things like this -> https://github.com/Eliav2/react-native-collapsible-view

File: src/components/TextInputWithName/index.js
Comment (marcaaron) [inline src/components/TextInputWithName/index.js:7]: `attribute`

File: src/pages/signin/SignInPage.js
Comment (marcaaron) [inline src/pages/signin/SignInPage.js:97]: After seeing this I think this utility would be a bit cleaner as an HOC. Passing the prop only to immediately consume it looks kind of funny.

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [reply src/pages/signin/PasswordForm.js:71]: Thoughts on how we might be able to do it? Or if it's worth doing?

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [reply src/pages/signin/LoginForm.js:77]: Still waiting for the comments / to see what you have in mind for this :)

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: src/components/CollapsibleView.js
Comment (parasharrajat) [reply src/components/CollapsibleView.js:18]: Yup, Thanks for the better name

File: src/pages/signin/SignInPage.js
Comment (parasharrajat) [reply src/pages/signin/SignInPage.js:97]: Hmm. Let me do it that way.

Comment (parasharrajat) [reply src/pages/signin/SignInPage.js:97]: Actually, prop is used to manage side effects. Ased on visibility.

File: src/pages/signin/PasswordForm.js
Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js:71]: One way or another we have to manually set the focus when page becomes visible. 
So what I can think of is that we pass a callback from the `ToggleVisibilityView` which we call here to set the ref of the input.
and Move the whole logic to focus the input in the HOC. 

we set the focus whenever page becomes visible and the ref is set.

so in this component we just need to call something like.
```this.props.setRefForToggleVisibility(this.textinput);```

File: src/pages/signin/SignInPage.js
Comment (marcaaron) [reply src/pages/signin/SignInPage.js:97]: Could maybe pass the prop through to the wrapped component?

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [reply src/pages/signin/PasswordForm.js:71]: Could forward the ref and do 

```jsx
<WrappedComponent
    ref={el => this.focusableElement = el}
/>
```

Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js:71]: I don't think it would be as straight as this. we will have to forwardRef the from LoginForm. so in future any component that is using this HOC has to manually forward the ref. 
Where calling this.props.setRefForToggleVisibility(this.textinput); would be much better.

`WrappedComponent` is not a Textinput. It could be group of elments and like a page. So forwarding a ref does not sound good. It is unclear ,whose ref will it be on the page etc.

Comment (marcaaron) [reply src/pages/signin/PasswordForm.js:71]: Hmm well there is only one "ref" that we are concerned about in the HOC right?

>so in future any component that is using this HOC has to manually forward the ref

Seems pretty much the same if you have to do `setRefForToggleVisibility(this.textinput)` - otherwise there will be no ref.

But maybe it is not very straightforward to do either - and in this case might just be better to be explicit. Want to pivot and just add some comments about why the logic is required? We can always revisit this later if the pattern pops up a lot.

Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js:71]: Overall, I think it is not that much helpful atm. We do need to use isVisible prop on both LoginForm and passwordForm.

**Comment by parasharrajat** [comment]
Ready.

**Comment by parasharrajat** [comment]
Ready for review.

File: src/components/withToggleVisibilityView.js
Comment (marcaaron) [inline src/components/withToggleVisibilityView.js:8]: typo here

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [inline src/pages/signin/LoginForm.js:68]: Can probably get rid of this comment. All we need to know is that we focus the input when the component mounts or visibility changes and the code explains it well enough.

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [inline src/pages/signin/LoginForm.js:53]: I think it does not need to be bound?

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [inline src/pages/signin/LoginForm.js:99]: Why call clear on this? Is `ExpensiTextInput` not a controlled component?

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [inline src/pages/signin/PasswordForm.js:65]: Remove.

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [reply src/pages/signin/PasswordForm.js:70]: K let's just leave this for now.

File: src/pages/signin/PasswordForm.js
Comment (marcaaron) [inline src/pages/signin/PasswordForm.js:162]: Why?

File: src/pages/signin/SignInPage.js
Comment (marcaaron) [inline src/pages/signin/SignInPage.js:90]: I think we need to say more. Can we please explain that the fields remain accessible if hidden but not if unmounted? Basically this should read as a warning that password managers will break without using `isVisible` which "hides" but does not "unmount". We don't want anyone to break this unintentionally.

**Review by marcaaron — CHANGES_REQUESTED** [review]
Looks good - just have a few more suggestions.

File: src/pages/signin/LoginForm.js
Comment (parasharrajat) [reply src/pages/signin/LoginForm.js:99]: Yeah. It's not. I just saw that we don't pass the value directly. `ExpensiTextInput` is using `value={this.value}`. Actually, we never faced a situation to make it controlled.

File: src/pages/signin/LoginForm.js
Comment (parasharrajat) [reply src/pages/signin/LoginForm.js:68]: ok.

File: src/pages/signin/SignInPage.js
Comment (parasharrajat) [reply src/pages/signin/SignInPage.js:90]: I'll try my best but really bad at explaining.

Comment (marcaaron) [reply src/pages/signin/SignInPage.js:90]: No worries, something like this works:

>LoginForm and PasswordForm must use the isVisible prop. This keeps them mounted, but visually hidden so that password managers can access the values. Conditionally rendering these components will break this feature.

File: src/pages/signin/PasswordForm.js
Comment (parasharrajat) [reply src/pages/signin/PasswordForm.js]: `ONYXKEYS.CREDENTIALS` could be null so it was causing a runtime issue in `ChangeExpensifyLoginLink`. Now, I have made it optional and passed a default value so we don't need it.

**Comment by parasharrajat** [comment]
Changes Done.

File: src/pages/signin/LoginForm.js
Comment (marcaaron) [reply src/pages/signin/LoginForm.js:98]: Oh hmm weird I don't get it. We are passing a value - is it used or not ? 🤔 

It seems like a bad habit to call `this.input.clear` every time we update a value passed to `ExpensifyTextInput` to be an empty string. Why do we need to do this?

**Review by marcaaron — DISMISSED** [review]
LGTM, I'm not too sure about that `this.input.clear` usage - feels like something we should not have to worry about.

File: src/pages/signin/LoginForm.js
Comment (parasharrajat) [reply src/pages/signin/LoginForm.js:98]: I wasn't using it before but when after recent changes, `ExpensifyTextInput` value is not clearing on updating the state variable login. But as soon as you blur it, values clear. So I think that using `this.value` instead of the state variable is causing this issue on `ExpensifyTextInput`. 
As far as I know, it's not affecting other parts of the app so I don't think that refactor is necessary. If we face this issue again then I would suggest making the `ExpensifyTextInput` controlled.

**Comment by parasharrajat** [comment]
@Jag96 This is ready for merge.

**Review by Jag96 — CHANGES_REQUESTED** [review]
Looks good and tests well, looks like just one copy-paste mistake

File: src/components/ExpensiTextInput/ExpensiTextInputLabel/index.native.js
Comment (parasharrajat) [reply src/components/ExpensiTextInput/ExpensiTextInputLabel/index.native.js:40]: Yeah, Nice catch.

Comment (parasharrajat) [reply src/components/ExpensiTextInput/ExpensiTextInputLabel/index.native.js]: Done.

**Review by Jag96 — APPROVED** [review]

**Comment by parasharrajat** [comment]
Unfortunately, found an issue with this PR. 

The keyboard does not close when you switch between the login and password pages on Android. As a side-effect, View is not correctly placed above the keyboard when the page changes. It happens due to `keyboardShouldPersistTaps="handled"
` prop. Previously unmounting the input was used to close the Keyboard.

I guess, we will have to manually blur the input when `isvisible` is false in LoginForm and PasswordForm. @Jag96 . I will spin up the PR tomorrow.

**Pull request merged**
